autotest/cpp/test_gdal.cpp

```diff
@@ -2462,6 +2462,155 @@ TEST_F(test_gdal, GDALBufferHasOnlyNoData)
+                  GDT_Int32, std::numeric_limits<double>::lowest()),
+              0);
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Int32, std::numeric_limits<double>::max()),
+              0);
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Int32, std::numeric_limits<int32_t>::lowest()),
+              std::numeric_limits<int32_t>::lowest() + 1);
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Int32, std::numeric_limits<int32_t>::max()),
+              std::numeric_limits<int32_t>::max() - 1);
+
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_UInt64, std::numeric_limits<double>::lowest()),
+              0);
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_UInt64, std::numeric_limits<double>::max()),
+              0);
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_UInt64, std::numeric_limits<uint64_t>::lowest()),
+              std::numeric_limits<uint64_t>::lowest() + 1);
+    // uin64_t max is not representable in double so we expect the next value to be returned
+    EXPECT_EQ(
+        GDALGetRasterNoDataReplacementValue(
+            GDT_UInt64,
+            static_cast<double>(std::numeric_limits<uint64_t>::max())),
+        std::nextafter(
+            static_cast<double>(std::numeric_limits<uint64_t>::max()), 0) -
+            1);
+
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Int64, std::numeric_limits<double>::lowest()),
+              0);
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Int64, std::numeric_limits<double>::max()),
+              0);
+    // in64_t max is not representable in double so we expect the next value to be returned
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Int64, std::numeric_limits<int64_t>::lowest()),
+              std::numeric_limits<int64_t>::lowest() + 1);
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Int64,
+                  static_cast<double>(std::numeric_limits<int64_t>::max())),
+              std::nextafter(
+                  static_cast<double>(std::numeric_limits<int64_t>::max()), 0) -
+                  1);
+
+    // Test floating point types
+
+    // out of range for float32
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Float32, std::numeric_limits<double>::lowest()),
+              0.0);
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Float32, std::numeric_limits<double>::max()),
+              0.0);
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Float32, std::numeric_limits<double>::infinity()),
+              0.0);
+    EXPECT_EQ(GDALGetRasterNoDataReplacementValue(
+                  GDT_Float32, -std::numeric_limits<double>::infinity()),
+              0.0);
+    // in range for float 32
```

Comment:
we'd need a test for in-range ``std::numeric_limits<float>::max()`` and ``std::numeric_limits<double>::lowest()``